core/src/duplicate_repair_status.rs

```diff
@@ -464,12 +454,6 @@ impl AncestorRequestStatus {
                 our version of the dead slot chains to the wrong fork!",
                     ancestor_slot, self.requested_mismatched_slot
                 );
-                earliest_erroring_ancestor = Some((
```

Comment:
After reviewing this case, I believe this is more in line with the comment and original intention.

Either the mismatch is before the missing slot (handled in previous case), or after in which case we should ignore the missing slot and keep searching. Previously we would stop at the missing slot and rely on replay to dump the entire fork including the mismatch.

Reply:
Does this mean we can entirely remove the `DuplicateAncestorDecision::EarliestAncestorNotFrozen(_)` from the enum since it's never going to be set anywhere?

Reply:
> or after in which case we should ignore the missing slot and keep searching
I'm guessing you're referring to the comment here:  https://github.com/solana-labs/solana/blame/adc3663feb585668c12721c74bc5991b7896a297/core/src/duplicate_repair_status.rs#L434

Hmm, so the case where we want to keep searching is, given an ancestor sample vector `A` from the cluster:

1. Slot `A[0]`  is missing
2. Slot `A[1]` which is a later slot is frozen and mismatched

So this can only happen if our version of slot `A[1]` chains to the wrong ancestor, i.e. it does not have `A[0]` as the parent?

In this case, the new behavior would be to keep searching for the first frozen mismatch?  But what if all the descendants are not frozen because we have the wrong ancestor tree?

Something like, let's say we have a dead slot `9` that we are running `AncestorHashes` on:

Our tree `1 2 3 9`
Ancestor hashes returns: `1, 3, 5, 9`

Reply:
In this case, if 9 is dead that means we must have a frozen hash for 3. And we would then see the hashes are different and dump 3.

> But what if all the descendants are not frozen because we have the wrong ancestor tree?
They have to be frozen for us to play the original requested slot. The only case this could happen is if the requested slot itself was the mismatch, in which case the normal dump & repair will fix it.

Reply:
Oh sorry bad example let me see if I understand with a better example where there are no overlapping slots.

Let's say we have:

Our tree: `1, 2, 3, 9`
Ancestor hashes returns: `1, 4, 5, 9`

In our tree, `9` is dead. We see:

1. Slot 1 matches
2. Slot 4 is missing. In the old code we set `earliest_erroring_ancestor = DuplicateAncestorDecision::EarliestAncestorNotFrozen`. In this new change we don't set this.
3. Slot 5 is missing, we continue
4.  Slot 9 is dead and not frozen, we exit the loop with `earliest_erroring_ancestor = None` still


Then:

1. In today's code, because `earliest_erroring_ancestor` is `Some` we set the slots to dump here: https://github.com/jbiseda/solana/blob/546c056a9a44c96fa99c41f03ccc991cbcb43db6/core/src/duplicate_repair_status.rs#L489-L491, which would dump slots `{5, 9}`
2. In this changed code, `earliest_erroring_ancestor = None` so we return https://github.com/jbiseda/solana/blob/546c056a9a44c96fa99c41f03ccc991cbcb43db6/core/src/duplicate_repair_status.rs#L497 which only dumps slot `9`, preventing us from dumping slot `5`, which is the intention?

Do we have a test case for this case?

Reply:
Would also be nice to extend this comment: https://github.com/solana-labs/solana/blob/3841e803d97f1dbccea5647ddf8d14bebb201c4a/core/src/duplicate_repair_status.rs#L444-L445 to cover this above case, because the mismatch does not need to happen immediately after the missing slot as suggested by the comment, and in such cases this part of the comment: 

```
Because we know any mismatches will also trigger the mismatch casing earlier in
                // the function, we will return`EarliestMismatchFound(first_mismatch)`.
```

also does not happen

Reply:
In the above case, `ancestor_hashes_service` would not return anything, as none of the ancestors are problematic, only slot 9 itself. 

If 9 itself is duplicate confirmed or epoch slots frozen, it will be dumped and repaired through normal means (`on_dead_slot` or `on_duplicate_confirmed` in `cluster_slot_state_verifier.rs`). This pathway is tested in the tests `duplicate_confirmed_state_update_1` , `dead_state_update_2`, and `dead_state_update_3`.

There is one other case that can happen, which is if for some reason 9 itself is never viewed as duplicate confirmed or initially epoch slots frozen. In that case we have to rely on the fact that one of its descendants will be viewed as such and run ancestor repair.

Reply:
In the old code, 9 will almost always be dumped and repaired through the state machine, and meanwhile 5, 9 will be dumped because of ancestor hashes. I think this approach is unnecessary, because there's no reason for 5 or any of the missing slots to be dumped. 

The functionality that we're missing is the following case:
- 9 arrives as dead (not duplicate confirmed or epoch slots frozen) -> sent to ancestor hashes
- later, 9 is found as epoch slots frozen by ancestor hashes, initiate request
- 5 and 9 are dumped
in which case with the current code, we would never dump 9.

For this case I will change the code such that if we have missing, and there is no frozen hash after, dump the requested slot

Reply:
Here's what i've done:
* Removed `AncestorsAllMatch`: if all the ancestors match, return the requested slot as the mismatch. I believe this was a bug with the previous implementation. We assumed that this was fine because normal `dump_then_repair` would take care of the slot, however the case i outlined above would fail to be detected, because the epoch slots frozen check happens only in `ancestor_hashes_service`. 
* This will result in double dumping, but without proper plumbing from the repairable pool to indicate how this slot was chosen for ancestor repair this is the best we can do. 
* The test `test_add_multiple_responses_ancestors_all_not_frozen` tests the case that you described above and only dumps the last slot (9 in your example).
* Updated the comments